src/abe/dippe.c

```diff
@@ -0,0 +1,428 @@
+
+    // C0 = g1^As (k+1 x 1)
+    cfe_mat_G1_mul_vec(&(cipher->C0), &(dippe->g1_A), &s);
+
+    // g1^(W_A_s) (k+1 x 1)
+    cfe_vec_G1 g1_W_A_s;
+    cfe_vec_G1_init(&g1_W_A_s, (dippe->assump_size + 1));
+
+    // g1^(x_UA_s) (k+1 x 1)
+    cfe_vec_G1 g1_x_UA_s;
+    cfe_vec_G1_init(&g1_x_UA_s, (dippe->assump_size + 1));
+
+    // Ci = g1^(xU(T)+W(T))As (pol x k+1)
+    for (size_t m = 0; m < pks_len; m++) {
+        // g1^(W(T)As)
+        cfe_mat_G1_mul_vec(&g1_W_A_s, &(pks[m]->g1_W_A), &s);
+
+        // g1^(xU(T)As)
+        for (size_t i = 0; i < (dippe->assump_size + 1); i++) {
+            ECP_BN254_inf(&(g1_x_UA_s.vec[i]));
+            for (size_t k = 0; k < dippe->assump_size; k++) {
+                ECP_BN254_copy(&tmp_g1, &(dippe->g1_UA.mat[i].vec[k]));
+                BIG_256_56_from_mpz(tmp_big, s.vec[k]);
+                ECP_BN254_mul(&tmp_g1, tmp_big);
+                ECP_BN254_add(&(g1_x_UA_s.vec[i]), &tmp_g1);
+            }
+            BIG_256_56_from_mpz(tmp_big, pv->vec[m]);
+            ECP_BN254_mul(&(g1_x_UA_s.vec[i]), tmp_big);
+        }
+
+        for (size_t i = 0; i < (dippe->assump_size + 1); i++) {
+            ECP_BN254_copy(&(cipher->Ci.mat[m].vec[i]), &(g1_x_UA_s.vec[i]));
+            ECP_BN254_add(&(cipher->Ci.mat[m].vec[i]), &(g1_W_A_s.vec[i]));
+        }
+    }
+
+    // C_prime = m * gt^alpha(T)As
+    FP12_BN254_one(&(cipher->C_prime));
+    for (size_t m = 0; m < pks_len; m++) {
+        for (size_t k = 0; k < dippe->assump_size; k++) {
+            BIG_256_56_from_mpz(tmp_big, s.vec[k]);
+            FP12_BN254_pow(&tmp_gt, &(pks[m]->gt_alpha_A.vec[k]), tmp_big);
+            FP12_BN254_mul(&(cipher->C_prime), &tmp_gt);
+        }
+    }
+    FP12_BN254_mul(&(cipher->C_prime), msg);
+
+    // cleanup
+    cfe_vec_free(&s);
+    cfe_vec_G1_free(&g1_W_A_s);
+    cfe_vec_G1_free(&g1_x_UA_s);
+
+    return CFE_ERR_NONE;
+}
+
+cfe_error cfe_dippe_keygen(cfe_dippe_user_sec_key *usk, cfe_dippe *dippe, size_t usk_id, cfe_dippe_pub_key *pks[], size_t pks_len, cfe_dippe_sec_key *sk, cfe_vec *av, char gid[]) {
+    // check whether there is the same number of keys as attribute vector components
+    if ((av->size != pks_len) || (pks_len <= usk_id)) {
+        return CFE_ERR_INSUFFICIENT_KEYS;
+    }
+
+    // temp vars
+    BIG_256_56 tmp_big;
+    ECP2_BN254 tmp_g2, hashed;
+    char *hash_str, *str_i;
+
+    // attribute vector as string
+    char *str_vec = cfe_vec_to_string(av);
+
+    // buffer for hashed group element
+    char str_ele[4 * MODBYTES_256_56];
+    octet oct;
+    oct.val = (char *)&str_ele;
+
+    // mue
+    ECP2_BN254 mue;
+    ECP2_BN254_inf(&mue);
+    for (size_t i = 0; i < pks_len; i++) {
+        ECP2_BN254_copy(&tmp_g2, &(pks[i]->g2_sigma));
+        BIG_256_56_from_mpz(tmp_big, sk->sigma);
+        ECP2_BN254_mul(&tmp_g2, tmp_big);
+
+        ECP2_BN254_toOctet(&oct, &tmp_g2);
+        hash_str = cfe_strings_concat(str_ele, "|", gid, "|", str_vec, NULL);
```

Comment:
Can't you just put `oct.val` instead of `str_ele` here?

Reply:
While changing this, I noticed that using the output of `ECP2_BN254_toOctet` directly with `cfe_strings_concat` doesn't work very well as it is an arbitrary byte array. I added another step to gain a properly null terminated string that can be concatenated. Please check again.